File: learnLeetCode/Day28/Part2/prime_factor_network.py
from typing import List

class DSU:
    def __init__(self, n: int):
        self.p = list(range(n))
        self.r = [0]*n
    def find(self, x):
        while self.p[x] != x:
            self.p[x] = self.p[self.p[x]]
            x = self.p[x]
        return x
    def union(self, a, b):
        ra, rb = self.find(a), self.find(b)
        if ra == rb: return False
        if self.r[ra] < self.r[rb]: ra, rb = rb, ra
        self.p[rb] = ra
        if self.r[ra] == self.r[rb]: self.r[ra] += 1
        return True

def factor(n: int) -> List[int]:
    res = []
    d = 2
    x = n
    while d*d <= x:
        if x % d == 0:
            res.append(d)
            while x % d == 0:
                x //= d
        d += 1 if d == 2 else 2  # after 2, test only odd numbers
    if x > 1:
        res.append(x)
    return res

def count_components(nums: List[int]) -> int:
    n = len(nums)
    dsu = DSU(n)
    prime_owner = {}
    for i, v in enumerate(nums):
        primes = factor(v)
        for p in primes:
            if p in prime_owner:
                dsu.union(i, prime_owner[p])
            else:
                prime_owner[p] = i
    roots = set(dsu.find(i) for i in range(n))
    return len(roots)

if __name__ == "__main__":
    nums = [2,3,6,15,35]
    print(count_components(nums))  # Expected 2

File: learnLeetCode_remote/Day28/Part2/prime_factor_network.cpp
#include <bits/stdc++.h>
using namespace std;

struct DSU{
    vector<int> p,r;
    DSU(int n):p(n),r(n,0){ iota(p.begin(), p.end(), 0); }
    int find(int x){ return p[x]==x?x:p[x]=find(p[x]); }
    bool unite(int a,int b){ a=find(a); b=find(b); if(a==b) return false; if(r[a]<r[b]) swap(a,b); p[b]=a; if(r[a]==r[b]) r[a]++; return true; }
};

vector<long long> factor(long long n){
    vector<long long> res; long long x=n; long long d=2;
    while(d*d<=x){
        if(x%d==0){ res.push_back(d); while(x%d==0) x/=d; }
        d += (d==2?1:2);
    }
    if(x>1) res.push_back(x);
    return res;
}

int count_components(const vector<long long>& nums){
    int n=nums.size(); DSU dsu(n); unordered_map<long long,int> owner; owner.reserve(n*2);
    for(int i=0;i<n;++i){
        auto primes=factor(nums[i]);
        for(auto p:primes){ auto it=owner.find(p); if(it==owner.end()) owner[p]=i; else dsu.unite(i,it->second); }
    }
    unordered_set<int> roots; roots.reserve(n*2);
    for(int i=0;i<n;++i) roots.insert(dsu.find(i));
    return (int)roots.size();
}

int main(){
    vector<long long> nums={2,3,6,15,35};
    cout<<count_components(nums)<<"\n"; // Expected 2
    return 0;
}
